### ISSUE TITLE
[Tracker] WebdriverIO v9 Release

### ISSUE THREAD

**Issue by christian-bromann** [webdriverio/webdriverio#10362]
This issue starts document upcoming (possibly breaking changes) in WebdriverIO v9. If you have any concerns about listed changes, please let us know. Also feel free to make suggestions to changes we should make in v9.

- Drop support for Node.js v16 and make v18 and v20 a requirement
- remove all dependencies for making request and replace it with `fetch` which landed as stable in v18

**Comment by nextlevelbeard** [comment]
I'd like to see https://github.com/webdriverio/webdriverio/issues/8149

We are using deprecated functions to run Cucumber and it is about time we upgrade.
Does not seem all that complex to substitute some deprecated functions with the new API.

**Comment by christian-bromann** [comment]
I am not sure if this has an impact to a WebdriverIO user because the integration to Cucumber is abstracted away. If you like to take a stab at this, please do.

**Comment by gpaciga** [comment]
@christian-bromann is there an eta for v9?

**Comment by christian-bromann** [comment]
no

**Comment by ichman** [comment]
Hello, as we're using the Devtools in our tests to monitor performance, will there be a way to still use it ?
Thank you

**Comment by christian-bromann** [comment]
> will there be a way to still use it ?

Yes, you will be able to use `getPuppeteer` to interact with Chrome DevTools.

**Comment by erwinheitzman** [comment]
Vite uses esbuild under the hood, I feel like we can use esbuild instead of Vite to transpile code.  What do you think?

**Comment by christian-bromann** [comment]
What would be the advantage of using esbuild directly over Vite?

**Comment by erwinheitzman** [comment]
> What would be the advantage of using esbuild directly over Vite?

I don't know what Vite would add that we need and since esbuild would cover our needs I feel like it should be sufficient to use that instead. Perhaps Vite offers features that we need or are very useful for our users but I do not know this so it's not like I think we should definitely do this but I think it's something to consider 👍

**Comment by christian-bromann** [comment]
My reasoning for choosing Vite over Esbuild were the following:

- Vite comes with a dev server that currently runs our component tests, esbuild is only a compiler and we would have to write all the logic to resolve files, compile them and cache them ourselves
- Vite is much easier to configure than esbuild
- its plugin system is much easier to wrap your head around and there are a lot of Vite plugins that not always work with esbuild only

Do you see Vite adding any cost in maintenance and performance? Overall it enables us to do a lot with almost no code.

**Comment by erwinheitzman** [comment]
Thank you for sharing that Christian! I have no objection or see any cost and have actually just read that Vite is working on replacing esbuild with rolldown, a Rust replacement that should achieve the same thing with better performance and minimal to no impact during the switch 👍

**Comment by christian-bromann** [comment]
> WebdriverIO 9 needs to support network intercepting

WebdriverIO already supports this in Chrome, see https://webdriver.io/docs/api/browser/mock. We plan to add cross browser support once required primitives are spec'ed out in WebDriver Bidi and landed in browser. That said, this type of feature can be added anytime independant of the upcoming release.

**Comment by goosewobbler** [comment]
Looking at picking up another one from this list - regarding the puppeteer task (no ticket), is that just adding it as a peerdep, since `getPuppeteer` [already checks](https://github.com/goosewobbler/webdriverio/blob/main/packages/webdriverio/src/commands/browser/getPuppeteer.ts#L48) for the presence of `puppeteer-core`?

**Comment by christian-bromann** [comment]
@goosewobbler that's awesome!

It seems like I already have done this in https://github.com/webdriverio/webdriverio/commit/220ad321925dff4b304e75056c3038b5f5b33e86

Would you be interested getting https://github.com/webdriverio/webdriverio/pull/12300 over the goal line? This essentially tries to rename the devtools service to google lighthouse service and removes all CDP integrations to only provide the lighthouse integration features. Wdyt?

**Comment by goosewobbler** [comment]
Yes, I can take a look at that one, also considered https://github.com/webdriverio/expect-webdriverio/issues/1366, is that about removing the matchers completely or is it just already covered by the deprecations?

**Comment by christian-bromann** [comment]
> also considered [webdriverio/expect-webdriverio#1366](https://github.com/webdriverio/expect-webdriverio/issues/1366),

Feel free to pick this up as well, the commands are already deprecated and only need to be removed. We may want to create a `@next` branch or something like that were we can push these changes for the upcoming major version.

**Comment by christian-bromann** [comment]
Some notes of breaking changes I've identified when migrating ecosystem packages to v9:

- the `webdriverio` packages doesn't export the `Mock` type anymore -> this has to be replaced with `WebdriverIO.Mock` as it is part of the WebdriverIO namespace now
- folks can't `await` a `ChainablePromiseElement` anymore and expect access to properties like `selector` or other things as the transformation from `ChainablePromiseElement` to `WebdriverIO.Element` is not supported anymore -> solution here is to call the helper methods `getElement` or `getElements`
- mock types have changed

**Comment by christian-bromann** [comment]
# Ecosystem Upgrade

or the v9 release, we aim to upgrade all ecosystem packages to v9 to ensure compatibility and identify any potential edge cases, bugs, or anomalies. Below is a list of pull requests for the ecosystem packages that have been upgraded to v9:

- https://github.com/webdriverio/expect-webdriverio/pull/1590

**Comment by christian-bromann** [comment]
We have released v9, I will go ahead and close this. There are some regression based on failures in some of the ecosystem projects. We will follow up with these issues individually. Thanks you all!

**Comment by shivam-mishra-deel** [comment]
Currently in WebdriverIO v9, mocking via browser.mock() only works when using the WebDriver Bidi protocol, and unfortunately:
WebDriver Bidi support is still experimental.
DevTools protocol (used by default in v9) doesn't support browser.mock() anymore.
Classic WebDriver (like Chromedriver + Selenium) also doesn't support browser.mock().
@christian-bromann

**Comment by christian-bromann** [comment]
> WebDriver Bidi support is still experimental.

Depends on your definition for experimental. It is available in Firefox and Chrome which accounts in the standard space for stable.

> DevTools protocol (used by default in v9) doesn't support browser.mock() anymore.

This allowed us to enable support for mocking on Firefox browser as well as we use WebDriver Bidi for it now.

> Classic WebDriver (like Chromedriver + Selenium) also doesn't support browser.mock().

Nope, you need WebDriver Bidi which is supported running Chromedriver + Selenium
